src/common/imageio_jpegxl.c

```diff
@@ -0,0 +1,242 @@
+    JxlResizableParallelRunnerDestroy(runner);
+    JxlDecoderDestroy(decoder);
+    free(read_buffer);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+  if(JxlDecoderSubscribeEvents(decoder, JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING | JXL_DEC_FULL_IMAGE) != JXL_DEC_SUCCESS)
+  {
+    fprintf(stderr, "[jpegxl_open] ERROR: JxlDecoderSubscribeEvents failed\n");
+    JxlResizableParallelRunnerDestroy(runner);
+    JxlDecoderDestroy(decoder);
+    free(read_buffer);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+  if(JxlDecoderSetParallelRunner(decoder, JxlResizableParallelRunner, runner) != JXL_DEC_SUCCESS)
+  {
+    fprintf(stderr, "[jpegxl_open] ERROR: JxlDecoderSetParallelRunner failed\n");
+    JxlResizableParallelRunnerDestroy(runner);
+    JxlDecoderDestroy(decoder);
+    free(read_buffer);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+
+  // Grand Decoding Loop
+  while(1)
+  {
+    status = JxlDecoderProcessInput(decoder);
+
+    if(status == JXL_DEC_ERROR)
+    {
+      fprintf(stderr, "[jpegxl_open] ERROR: JXL decoding failed\n");
+      JxlResizableParallelRunnerDestroy(runner);
+      JxlDecoderDestroy(decoder);
+      free(read_buffer);
+      return DT_IMAGEIO_FILE_CORRUPTED;
+    }
+
+    if(status == JXL_DEC_NEED_MORE_INPUT)
+    {
+      fprintf(stderr, "[jpegxl_open] ERROR: JXL data incomplete\n");
+      JxlResizableParallelRunnerDestroy(runner);
+      JxlDecoderDestroy(decoder);
+      free(read_buffer);
+      return DT_IMAGEIO_FILE_CORRUPTED;
+    }
+
+    if(status == JXL_DEC_BASIC_INFO)
+    {
+      if(JxlDecoderGetBasicInfo(decoder, &basicinfo) != JXL_DEC_SUCCESS)
+      {
+        fprintf(stderr, "[jpegxl_open] ERROR: JXL basic info not available\n");
+        JxlResizableParallelRunnerDestroy(runner);
+        JxlDecoderDestroy(decoder);
+        free(read_buffer);
+        return DT_IMAGEIO_FILE_CORRUPTED;
+      }
+
+      // Unlikely to happen, but let there be a sanity check
+      if(basicinfo.xsize == 0 || basicinfo.ysize == 0)
+      {
+        fprintf(stderr,"[jpegxl_open] ERROR: JXL image declares zero dimensions\n");
+        JxlResizableParallelRunnerDestroy(runner);
+        JxlDecoderDestroy(decoder);
+        free(read_buffer);
+        return DT_IMAGEIO_FILE_CORRUPTED;
+      }
+
+
+      num_threads = JxlResizableParallelRunnerSuggestThreads(basicinfo.xsize, basicinfo.ysize);
+      JxlResizableParallelRunnerSetThreads(runner, num_threads);
+
+      continue;    // go to next loop iteration to process rest of the input
+    }
+
+    if(status == JXL_DEC_COLOR_ENCODING)
+    {
+      if(JxlDecoderGetICCProfileSize(decoder, &pixel_format, JXL_COLOR_PROFILE_TARGET_DATA, &icc_size) == JXL_DEC_SUCCESS)
+      {
+        if(icc_size)
+        {
+          img->profile_size = icc_size;
+          img->profile = (uint8_t *)g_malloc0(icc_size);
+          JxlDecoderGetColorAsICCProfile(decoder, &pixel_format, JXL_COLOR_PROFILE_TARGET_DATA, img->profile, icc_size);
```

Comment:
But `if(icc_size)` checks that we will get the ICC profile successfully because icc_size is the size
of the profile we will get using the following JxlDecoderGetColorAsICCProfile function.

On the contrary, this check is actually redundant here, BTW, since the [documentation](https://libjxl.readthedocs.io/en/latest/api_decoder.html#_CPPv427JxlDecoderGetICCProfileSizePK10JxlDecoderPK14JxlPixelFormat21JxlColorProfileTargetP6size_t) guarantees that if the JxlDecoderGetICCProfileSize function returns JXL_DEC_SUCCESS, then the ICC profile is available.

Reply:
Ok